feat(curriculum): add binary search tree (#52371)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-tree-traversal-by-building-a-binary-search-tree/655a153a6b362d103e125028.md

---
id: 655a153a6b362d103e125028
title: Step 16
challengeType: 20
dashedName: step-16
---

# --description--

> If the condition where `key` < `node.key` is true, then the new node should be placed in the left subtree.

Reviewer: ```suggestion
If `key` < `node.key` is true, then the new node should be placed in the left subtree.
```

> Recursively call the `_insert` method on the left child of the current node. Then, assign the result to the left attribute of the current node.

Reviewer: ```suggestion
Recursively call the `_insert` method on the left child of the current node. Then, assign the result to the `left` attribute of the current node.
```

# --hints--

> Call the self._insert method using the parameters `node.left` and `key`.

Reviewer: ```suggestion
Call the `_insert` method passing `node.left` and `key` as the parameters. Remember to use `self`.
```

```js
({ test: () => assert.match(code, /self\._insert\(node\.left,\s*key\)/) });
```

Assign the result to the left attribute (`node.left`)of the current node.

```js
({ test: () => assert.match(code, /node\.left\s*=\s*self\._insert\(node\.left,\s*key\)/) });
```

# --seed--

## --seed-contents--

```py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self,key):
        self.root = self._insert(self.root, key)

>     def _insert(self, node, key):
>         if node is None:
>             return TreeNode(key)
>         if key < node.key:
> --fcc-editable-region--
>             
> --fcc-editable-region--

Reviewer: ```suggestion
--fcc-editable-region--
    def _insert(self, node, key):
        if node is None:
            return TreeNode(key)
        if key < node.key:            
--fcc-editable-region--
```
